### ISSUE TITLE
log1p(x), logb(x), ilogb(x) are not folded at the compilation time when x is constant

### ISSUE THREAD

**Issue by k-arrows** [llvm/llvm-project#101873]
Consider the following C++ example. The functions log2(x), log10(x) and log(x) are folded, but log1p(x), logb(x), ilogb(x) are not.
https://godbolt.org/z/nzrncEE37
```cpp
#include <cmath>

double f1() {
  double num = 8.0;
  return std::log2(num);
}

double f2() {
  double num = 100.0;
  return std::log10(num);
}

double f3() {
  double num = 2.718281828459045;
  return std::log(num);
}

double f4() {
  double num = 2.718281828459045;
  return std::log1p(num - 1);
}

double f5() {
  double num = 48.0;
  return std::logb(num);
}

int f6() {
  double num = 48.0;
  return std::ilogb(num);
}
```

**Comment by tharunvk** [comment]
Can you please assign this issue to me? I'm working on it.

**Comment by fawdlstty** [comment]
@tharunvk Hi, are you still work on it?

**Comment by tharunvk** [comment]
Yes. I'm done with logb and log1p. Almost done with ilogb. I'll raise the pull request soon

**Comment by badumbatish** [comment]
happy to give this a try if you're busy @tharunvk

**Comment by kshitijvp** [comment]
I checked for log1px and logbx it already works
But for ilogbx the call instruction is not removed so has to be optimized 


IR:
"define dso_local noundef i32 @_Z2f6v() local_unnamed_addr #0 {
entry:
  %call = tail call i32 @ilogb(double noundef 4.800000e+01) #2, !tbaa !5
  ret i32 5
}"

**Comment by VedantParanjape** [comment]
> Maybe some extra work is needed here:
> 
> [llvm-project/llvm/lib/Analysis/ConstantFolding.cpp](https://github.com/llvm/llvm-project/blob/5f1f69368e1f17ff9e32428f36cf34ffb6255ae5/llvm/lib/Analysis/ConstantFolding.cpp#L3520)
> 
> Line 3520 in [5f1f693](/llvm/llvm-project/commit/5f1f69368e1f17ff9e32428f36cf34ffb6255ae5)
>  bool llvm::isMathLibCallNoop(const CallBase *Call,

I doubt it's there, it seems like a bug in constant folding as the call is not being eliminated.

**Comment by VedantParanjape** [comment]
> Maybe some extra work is needed here:
> 
> [llvm-project/llvm/lib/Analysis/ConstantFolding.cpp](https://github.com/llvm/llvm-project/blob/5f1f69368e1f17ff9e32428f36cf34ffb6255ae5/llvm/lib/Analysis/ConstantFolding.cpp#L3520)
> 
> Line 3520 in [5f1f693](/llvm/llvm-project/commit/5f1f69368e1f17ff9e32428f36cf34ffb6255ae5)
>  bool llvm::isMathLibCallNoop(const CallBase *Call,

EarlyCSE seems to be doing the elimination.
